stan/math/opencl/cache_copy.hpp

```diff
@@ -0,0 +1,51 @@
+#ifndef STAN_MATH_OPENCL_CACHE_COPY_HPP
+#define STAN_MATH_OPENCL_CACHE_COPY_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/opencl_context.hpp>
+#include <stan/math/opencl/kernel_cl.hpp>
+#include <stan/math/prim/mat/fun/Eigen.hpp>
+#include <CL/cl.hpp>
+
+namespace stan {
+namespace math {
+namespace internal {
+template <int R, int C>
+inline void cache_copy(cl::Buffer dst, const Eigen::Matrix<double, R, C>& src) {
+  cl::Context& ctx = opencl_context.context();
+  cl::CommandQueue queue = opencl_context.queue();
+#ifdef STAN_OPENCL_CACHE
```

Comment:
do we ever not want the cache on if STAN_OPENCL is defined? I might have said something that generated this, if so I am sorry and my mistake!

Reply:
The issue here is with older GPUs with not much global memory, caching in that case isnt really useful as it fills up that memory. For current GPUs this isnt really an issue. So compatibility and trading memory space for speed.

Reply:
ah ok